waiter/src/waiter/metrics_consumer.clj

```diff
@@ -0,0 +1,196 @@
+            [waiter.metrics :as metrics]
+            [waiter.service-description :as sd]
+            [waiter.token :as token]
+            [waiter.util.http-utils :as hu]
+            [waiter.util.utils :as utils]))
+
+(defn make-metrics-watch-request
+  "Creates a watch request to a metrics service and pushes token events to the provided event-chan. If the watch request
+  fails for any reason, it will retry after the retry-delay-ms. The metrics service needs to stream chunked json on the
+  /token-stats?watch=true endpoint. Each json blob has the structure:
+  {:object [
+    {:token \"token\"
+     :lastRequestTime iso-time-string}
+    ...]
+   :type \"initial\" or \"update\"}
+
+  A map is returned with keys:
+  :exit-chan is a channel that when a message is pushed to, terminates the watch immediately and go-chan
+  :go-chan channel closed when watch is terminated (used primarily for cleaning up tests)
+  :query-state-fn function called with include flags for introspecting the state of the watch (e.g. last-event-time)"
+  [http-client http-streaming-request-async-fn router-id cur-cid {:keys [cluster url] :as metrics-service-conf}
+   event-chan clock retry-delay-ms]
+  (cid/with-correlation-id
+    (str cur-cid ".metrics-watch." cluster)
+    (let [_ (log/info "starting metrics-service-watch" {:metrics-service-conf metrics-service-conf})
+          exit-chan (async/promise-chan)
+          exit-chan-mult (async/mult exit-chan)
+          state-atom (atom {:last-event-time nil
+                            :last-watch-time nil})
+          metrics-endpoint (str url "/token-stats")
+          query-string (str "watch=true&watcher=" router-id)
+          query-state-fn
+          (fn query-metrics-watch-fn
+            [_]
+            (let [{:keys [last-event-time last-watch-time]} @state-atom]
+              {:endpoint metrics-endpoint
+               :last-event-time last-event-time
+               :last-watch-time last-watch-time
+               :supported-include-params []}))
+          go-chan
+          (async/go-loop []
+            (try
+              (log/info "starting watch on metrics service:" {:metrics-endpoint metrics-endpoint})
+              (let [abort-ch (async/promise-chan)
+                    {:keys [body-chan error-chan] :as res}
+                    (async/<! (http-streaming-request-async-fn http-client metrics-endpoint
+                                                               :abort-ch abort-ch
+                                                               :query-string query-string))]
+                ; when async chan is triggered, we want to also trigger the abort channel on the watch connection
+                (async/tap exit-chan-mult abort-ch)
+                (when (instance? Throwable res)
+                  (throw res))
+                (swap! state-atom assoc :last-watch-time (clock))
+                (doseq [{:strs [object type]} (utils/chan-to-json-seq!! body-chan)]
```

Comment:
Is the blocking call `utils/chan-to-json-seq!!` inside a go-block intentional?

Reply:
its because its calling chan-to-seq!!, I couldn't find a nonblocking way of creating an InputStreamReader. from a channel. But this chan-to-seq!! should not block the caller as it is all lazy-sequences

Reply:
I think likely we'll need to override the InputStreamReader. with `proxy` and implement the `read` functions to use the channel directly. I can do this in a followup pr as well if that is okay with you

Reply:
The caller is blocked in the doseq when trying to read the next element.